File: Stack And Queue/225.用队列实现栈.cpp
#include <queue>
using namespace std;
/*
 * @lc app=leetcode.cn id=225 lang=cpp
 *
 * [225] 用队列实现栈
 */

// @lc code=startclass MyStack
class MyStack
{
public:
    queue<int> qStart;
    queue<int> qEnd;
    /** Initialize your data structure here. */
    MyStack()
    {
    }

    /** Push element x onto stack. */
    void push(int x)
    {
        qStart.push(x);
    }

    /** Removes the element on top of the stack and returns that element. */
    int pop()
    {
        int size = qStart.size();

        // 遍历qStart，直到找到他的最后一个，最后一个不进行pop去除，直接返回
        while (size > 1)
        {
            qEnd.push(qStart.front());
            qStart.pop();
            size--;
        }

        int result = qStart.front();
        qStart = qEnd; // 重新将新队列赋值给qStart

        // 清空que2
        while (!qEnd.empty())
        {
            qEnd.pop();
        }
        return result;
    }

    /** Get the top element. */
    int top()
    {
        return qStart.back();
    }

    /** Returns whether the stack is empty. */
    bool empty()
    {
        return qStart.empty();
    }
};
/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */
// @lc code=end

File: Stack And Queue/225.用队列实现栈.ts
/*
 * @lc app=leetcode.cn id=225 lang=typescript
 *
 * [225] 用队列实现栈
 */

// @lc code=start
class MyStack {
    arr: any
    constructor() {
        this.arr = []
    }
    // 将元素 x 压入栈顶
    push(x: number): void {
        // 放在数组头部
        this.arr = [x].concat(this.arr)
    }
    // 移除并返回栈顶元素
    pop(): number {
        // 返回头部元素
        return this.arr.shift()
    }
    // 返回栈顶元素
    top(): number {
        if (this.empty()) {
            return null
        }

        return this.arr[0]
    }
    // 如果栈是空的，返回 true ；否则，返回 false 。
    empty(): boolean {
        return this.arr.length === 0
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */
// @lc code=end

